File: askmate/rag.py
from __future__ import annotations
from typing import Dict, List

from askmate.config import AppConfig
from askmate.llm import generate_answer
from askmate.store import AskMateVectorStore


SYSTEM_PROMPT = (
    "You are AskMate, a precise and helpful AI knowledge buddy for teams. "
    "Prefer information from the provided context. Provide concise answers with bullet points when helpful. "
    "Cite sources by including their titles/links if provided in the context."
)


def build_user_prompt(query: str, context_chunks: List[Dict]) -> str:
    context_text = "\n\n".join([f"[Source {i+1}] {c['metadata'].get('source','')}\n{c['text']}" for i, c in enumerate(context_chunks)])
    return f"Question: {query}\n\nContext:\n{context_text}\n\nAnswer:"


def answer_question(query: str, vector_store: AskMateVectorStore, config: AppConfig, idk_mode: bool = True, top_k: int = 6, temperature: float = 0.2, channel: str = "web") -> Dict:
    # Retrieve
    context_chunks = vector_store.similarity_search(query, k=top_k)
    from_docs = len(context_chunks) > 0

    # Build prompt
    user_prompt = build_user_prompt(query, context_chunks)
    answer = generate_answer(SYSTEM_PROMPT, user_prompt, config, temperature)

    sources: List[str] = []
    for c in context_chunks:
        src = c["metadata"].get("source") or c["metadata"].get("title") or c["metadata"].get("url")
        if src and src not in sources:
            sources.append(src)

    # IDK mode: if no docs and short answer, flag missing
    if idk_mode and not from_docs:
        return {
            "answer": "I couldn't find this in your docs. Do you want me to flag this as a missing FAQ?",
            "sources": [],
            "from_docs": False,
            "suggest_flag": True,
        }

    return {
        "answer": answer,
        "sources": sources,
        "from_docs": from_docs,
        "suggest_flag": False,
    }

File: node-backend/src/rag.ts
import { VectorStore } from './vectorStore';

export async function answerQuestion(params: {
  query: string;
  store: VectorStore;
  idkMode: boolean;
  topK: number;
  temperature: number;
}): Promise<{
  answer: string;
  sources: string[];
  fromDocs: boolean;
  suggestFlag: boolean;
}> {
  const { query, store, idkMode, topK, temperature } = params;

  const chunks = await store.similaritySearch(query, topK);
  const fromDocs = chunks.length > 0;
  const context = chunks
    .map((c, i) => `[Source ${i + 1}] ${c.metadata?.source || ''}\n${c.text}`)
    .join('\n\n');

  if (idkMode && !fromDocs) {
    return {
      answer:
        "I couldn't find this in your docs. Do you want me to flag this as a missing FAQ?",
      sources: [],
      fromDocs: false,
      suggestFlag: true,
    };
  }

  const system =
    'You are AskMate, a precise and helpful AI knowledge buddy for teams. Prefer info from provided context. Cite sources by title/link.';
  const user = `Question: ${query}\n\nContext:\n${context}\n\nAnswer:`;

  const answer = await generate(system, user, temperature);

  const sources: string[] = [];
  for (const c of chunks) {
    const src = c.metadata?.source || c.metadata?.title || c.metadata?.url;
    if (src && !sources.includes(src)) sources.push(src);
  }

  return { answer, sources, fromDocs, suggestFlag: false };
}

async function generate(system: string, user: string, temperature: number): Promise<string> {
  const provider =
    (process.env.LLM_PROVIDER ||
      process.env.EMBEDDING_PROVIDER ||
      'gemini')
      .toLowerCase();

  if (provider === 'openai') {
    const key = process.env.OPENAI_API_KEY;
    if (!key) throw new Error('OPENAI_API_KEY is not set');
    try {
      const OpenAI = (await import('openai')).default;
      const client = new OpenAI({ apiKey: key });
      const model = process.env.OPENAI_CHAT_MODEL || 'gpt-4o-mini';
      const resp = await client.chat.completions.create({
        model,
        temperature,
        messages: [
          { role: 'system', content: system },
          { role: 'user', content: user },
        ],
      });
      return resp.choices?.[0]?.message?.content || '';
    } catch (err) {
      console.error('OpenAI error:', err);
      throw new Error('OpenAI model call failed');
    }
  }

  if (provider === 'gemini') {
    const key = process.env.GEMINI_API_KEY;
    if (!key) throw new Error('GEMINI_API_KEY is not set');
    try {
      const { GoogleGenerativeAI } = await import('@google/generative-ai');
      const genai = new GoogleGenerativeAI(key);
      const modelName = process.env.GEMINI_CHAT_MODEL || 'gemini-1.5-flash';
      const model = genai.getGenerativeModel({ model: modelName });

      const r = await model.generateContent({
        contents: [
          {
            role: 'user',
            parts: [{ text: system }, { text: user }],
          },
        ],
        generationConfig: { temperature },
      } as any);

      return r.response.text();
    } catch (err) {
      console.error('Gemini error:', err);
      throw new Error('Gemini model call failed');
    }
  }

  throw new Error(
    `Unsupported provider "${provider}". Set LLM_PROVIDER to "gemini" or "openai".`
  );
}
